### PR TITLE
[RPD-304] Migrate existing Terraform code into a modular format

### PR DESCRIPTION
**Pull request by Christopher-Norman**
This PR creates a new folder `src/matcha_ml/infrastructure/modules/` which contains all the Terraform code split into modules in preparation for the modular stack implementation. The 'common' module folder is a requirement/dependency for all (or most) other modules and contains AKS, ACR etc. 

The modules are in such a format that they can easily be copied and concatenated together as one cohesive Terraform deployment.

TODO:
1. Figure best way to set variables (not all variables will exist in every stack)

## Checklist

Please ensure you have done the following:

* [x] I have read the [CONTRIBUTING](https://github.com/fuzzylabs/matcha/blob/main/CONTRIBUTING.md) guide.
* [ ] I have updated the documentation if required.
* [ ] I have added tests which cover my changes.

## Type of change

Tick all those that apply:

* [ ] Bug Fix (non-breaking change, fixing an issue)
* [x] New feature (non-breaking change to add functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Other (add details above)

### CODE REVIEW

File: src/matcha_ml/infrastructure/modules/zenml/zen_server/ingress.tf
Comment (Christopher-Norman) [inline src/matcha_ml/infrastructure/modules/zenml/zen_server/ingress.tf:1]: Remove?

File: src/matcha_ml/infrastructure/modules/mlflow/mlflow_module/zenml_namespace.tf
Comment (Christopher-Norman) [inline src/matcha_ml/infrastructure/modules/mlflow/mlflow_module/zenml_namespace.tf:1]: Is this required?

Comment (Christopher-Norman) [reply src/matcha_ml/infrastructure/modules/mlflow/mlflow_module/zenml_namespace.tf:1]: Why is the `zenml_namespace` in the `mlflow` module?

Comment (d-lowl) [reply src/matcha_ml/infrastructure/modules/mlflow/mlflow_module/zenml_namespace.tf:1]: I think it's needed by the zenserver, but yeah, should be moved

File: src/matcha_ml/infrastructure/modules/mlflow/mlflow_module/outputs.tf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+output "mlflow_tracking_url" {
+  description = "The tracking URL for MLFlow dashboard"
+  value = "http://${data.kubernetes_service.mlflow_tracking.status.0.load_balancer.0.ingress.0.ip}:${data.kubernetes_service.mlflow_tracking.spec.0.port.0.port}"
```
Comment (d-lowl) [inline src/matcha_ml/infrastructure/modules/mlflow/mlflow_module/outputs.tf:3]: Is it the same ingress that's used by seldon? If so, it won't be there, if seldon's not deployed. Potentially needs moving around

File: src/matcha_ml/infrastructure/modules/seldon/seldon/istio.tf [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Derived from ZenML's stack recipes; source: https://github.com/zenml-io/mlops-stacks/blob/8eb06596bf836d3a3dd2634fbc7f2b5687421811/aws-minimal/seldon/istio.tf
+
+# create a namespace for istio resources
+resource "kubernetes_namespace" "istio_ns" {
```
Comment (d-lowl) [inline src/matcha_ml/infrastructure/modules/seldon/seldon/istio.tf:4]: Is istio only used by Seldon? If not, this needs to be in common module

File: src/matcha_ml/infrastructure/modules/zenml/zen_server/ingress.tf
Comment (d-lowl) [reply src/matcha_ml/infrastructure/modules/zenml/zen_server/ingress.tf:1]: At seems that nginx is not used indeed

File: src/matcha_ml/infrastructure/modules/zenml/zen_server/main.tf [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  set {
+    name  = "zenml.deploymentType"
+    value = "azure"
+  }
+  set {
+    name  = "zenml.analyticsOptIn"
+    value = var.analytics_opt_in
+  }
+
+  # # Ingress set up
```
Comment (d-lowl) [inline src/matcha_ml/infrastructure/modules/zenml/zen_server/main.tf:38]: Same here. This has been commented out for a while

File: src/matcha_ml/infrastructure/modules/zenml/zen_server/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  value = var.username
+}
+
+output "zenserver_password" {
+  description = "The password used to access the ZenML server"
+  value     = var.password
+  sensitive = true
+}
+
+# output "zenserver_url" {
```
Comment (d-lowl) [inline src/matcha_ml/infrastructure/modules/zenml/zen_server/outputs.tf:17]: Oh actually, wasn't it a thing that we wanted to fix?

**Review by d-lowl — COMMENTED** [review]
Generally looks good. I would give an additional thought to what settings we want to expose as variables in those modules. I've left some comments

**Review by JonoCX — CHANGES_REQUESTED** [review]

File: src/matcha_ml/infrastructure/modules/chroma/chroma/chroma_helm/Chart.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v2
+name: chroma
+description: Chroma Server Helm Chart
```
[the remark this answers is not in the document]
Comment (Christopher-Norman) [reply src/matcha_ml/infrastructure/modules/chroma/chroma/chroma_helm/Chart.yaml:3]: Not officially by Chroma DB, no.

File: src/matcha_ml/infrastructure/modules/seldon/seldon/permissions.tf
Comment (Christopher-Norman) [inline src/matcha_ml/infrastructure/modules/seldon/seldon/permissions.tf:1]: This file is only required if ZenML is deployed

File: src/matcha_ml/infrastructure/modules/chroma/chroma/chroma_helm/Chart.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (archena) [reply src/matcha_ml/infrastructure/modules/chroma/chroma/chroma_helm/Chart.yaml:3]: Chris and I have discussed possibly releasing our own in the near future FYI

File: src/matcha_ml/infrastructure/modules/seldon/seldon/istio.tf (hunk 2 above)
Comment (Christopher-Norman) [reply src/matcha_ml/infrastructure/modules/seldon/seldon/istio.tf:4]: It seems it is only used by Seldon

**Review by JonoCX — APPROVED** [review]

**Pull request merged**